src/main/java/com/openlattice/datastore/data/controllers/DataController.java

```diff
@@ -1025,31 +1011,46 @@ public long getEntitySetSize( @PathVariable( ENTITY_SET_ID ) UUID entitySetId )
             path = "/" + SET_ID_PATH + "/" + ENTITY_KEY_ID_PATH + "/" + PROPERTY_TYPE_ID_PATH,
             produces = MediaType.APPLICATION_JSON_VALUE )
     public Set<Object> getEntity(
-            @PathVariable( SET_ID_PATH ) UUID entitySetId,
+            @PathVariable( ENTITY_SET_ID ) UUID entitySetId,
             @PathVariable( ENTITY_KEY_ID ) UUID entityKeyId,
             @PathVariable( PROPERTY_TYPE_ID ) UUID propertyTypeId ) {
         ensureReadAccess( new AclKey( entitySetId ) );
-        EntitySet es = edmService.getEntitySet( entitySetId );
+        final EntitySet entitySet = edmService.getEntitySet( entitySetId );
 
-        if ( es.isLinking() ) {
-            checkState( !es.getLinkedEntitySets().isEmpty(),
+        if ( entitySet.isLinking() ) {
+            checkState( !entitySet.getLinkedEntitySets().isEmpty(),
                     "Linked entity sets are empty for linking entity set %s", entitySetId );
 
-            Map<UUID, Map<UUID, PropertyType>> authorizedPropertyTypes = authzHelper
-                    .getAuthorizedPropertyTypes( es.getLinkedEntitySets(),
+            entitySet.getLinkedEntitySets().forEach( esId -> ensureReadAccess( new AclKey( esId ) ) );
+
+            final Map<UUID, Map<UUID, PropertyType>> authorizedPropertyTypes = authzHelper
+                    .getAuthorizedPropertyTypesByNormalEntitySet(
+                            entitySet,
                             Set.of( propertyTypeId ),
                             EnumSet.of( Permission.READ ) );
-            FullQualifiedName propertyTypeFqn = authorizedPropertyTypes.get( entitySetId ).get( propertyTypeId )
+
+            // if any of its normal entitysets don't have read permission on property type, reading is not allowed
+            if ( authorizedPropertyTypes.values().iterator().next().isEmpty() ) {
```

Comment:
shouldn't this be `authorizedPropertyTypes.values().iterator().next().containsKey( propertyTypeId )` ?

Reply:
The getAuthorizedPropertyTypesByNormalEntitySet takes the property type as argument, so, if the calling user has no access, it should be empty, but that would work too.